python/cucim/src/cucim/skimage/_shared/distance.py

```diff
@@ -0,0 +1,177 @@
+    very wasteful if we  only want to find the maximum pointwise distance. This
+    function processes smaller blocks of the overall cdist result, keeping
+    track of the maximum value (and corrdsponding indices) across all blocks.
+
+    A schematic of the block sizes processed for an array of 10,000 coordinates
+    with a block size of 4000 points is shown below:
+
+      ┌────────┬────────┬────────┐
+      │  4000  │  4000  │  2000  │
+      │  4000  │  4000  │  2000  │
+      ├────────┼────────┼────────┤
+      │   X    │  4000  │  2000  │
+      │        │  4000  │  2000  │
+      ├────────┼────────┼────────┤
+      │   X    │   X    │  2000  │
+      │        │        │  2000  │
+      └────────┴────────┴────────┘
+
+    where due to the symmetry of the pdist matrix, we don't compute the lower
+    triangular blocks.
+    """
+
+    if _distance_on_cpu:
+        warnings.warn(
+            "cuVS >= 25.02 or pylibraft < 24.12 must be installed to use "
+            "GPU-accelerated pairwaise distance computations. Falling back "
+            "to SciPy-based CPU implementation."
+        )
+        xp = np
+        coords = cp.asnumpy(coords)
+    else:
+        xp = cp
+        coords = cp.asarray(coords)
+
+    num_coords, _ = coords.shape
+    if num_coords == 0:
+        raise RuntimeError("No coordinates to process")
+
+    if "out" in cdist_kwargs:
+        raise ValueError(
+            "'out' cannot be provided via cdist_kwargs (reserved "
+            "for internal use)"
+        )
+
+    blocks_per_dim = math.ceil(num_coords / coords_per_block)
+    if blocks_per_dim > 1:
+        # reuse the same temporary storage array for most blocks
+        # (last block in row and column may be smaller)
+        temp = xp.zeros((coords_per_block, coords_per_block), dtype=xp.float32)
+    if coords.dtype not in [xp.float32, xp.float64]:
+        coords = coords.astype(xp.float32, copy=False)
+    if not coords.flags.c_contiguous:
+        coords = xp.ascontiguousarray(coords)
+    max_dist = 0
+    for i in range(blocks_per_dim):
+        for j in range(blocks_per_dim):
+            if j < i:
+                # skip symmetric regions
+                continue
+            sl_m = slice(
+                i * coords_per_block,
+                min((i + 1) * coords_per_block, num_coords),
+            )
+            sl_n = slice(
+                j * coords_per_block,
+                min((j + 1) * coords_per_block, num_coords),
+            )
+            coords_block1 = coords[sl_m, :]
+            coords_block2 = coords[sl_n, :]
+            if i < blocks_per_dim - 1 and j < blocks_per_dim - 1:
+                distance.cdist(
+                    coords_block1,
+                    coords_block2,
+                    metric=metric,
+                    out=temp,
+                    **cdist_kwargs,
+                )
+                current_output = temp
+            else:
+                # omit out= for the last block as size may be
```

Comment:
This comment sentence doesn't seem to be complete.

Reply:
right, it is supposed to say "size may be different". I will update it